res/controllers/Traktor-Kontrol-S4-MK3.js

```diff
@@ -1444,69 +1444,132 @@ let wheelTimerDelta = 0;
-                } else if (pressed) {
-                    if (this.unit.focusedEffect !== null) {
-                        this.unit.setFocusedEffect(null);
-                    } else {
-                        script.toggleControl(this.unit.group, "group_[Master]_enable");
-                        this.shift();
+                    // Switch effect unit
+                    let newIndex = this.unit.unitNumbers.indexOf(this.unit.unitNumber) + 1;
+                    if (newIndex === (this.unit.unitNumbers.length)) {
+                        newIndex = 0; // wrap around
                     }
+                    const newUnitNumber = this.unit.unitNumbers[newIndex];
+                    // Make sure we have 4 units on screen when we switch to unit 3/4.
+                    if (newUnitNumber > 2) {
+                        engine.setValue("[Skin]", "show_4effectunits", 1);
+                    }
+                    this.unit.setCurrentUnit(newUnitNumber);
+                } else if (this.unit.focusedEffect !== null) {
+                    this.unit.setFocusedEffect(null);
+                    this.unit.stopFocusSelectMode();
+                } else {
+                    script.toggleControl(this.group, "show_parameters");
                 }
+                this.output(0);
+            },
+            onLongPress: function() {
+                if (this.shifted) {
+                    script.toggleControl(this.group, "show_parameters");
+                } else {
+                    this.unit.startFocusSelectMode();
+                }
+            },
+            onLongRelease: function() {
+                this.unit.stopFocusSelectMode();
+                this.output(0);
             }
         });
 
         this.knobs = [];
         this.buttons = [];
         for (const index of [0, 1, 2]) {
-            const effectGroup = `[EffectRack1_EffectUnit${unitNumber}_Effect${index + 1}]`;
+            const effectGroup = this.effectGroupForNumber(index + 1);
             this.knobs[index] = new Pot({
+                unit: this,
                 inKey: "meta",
                 group: effectGroup,
                 inReport: inReports[2],
                 inByte: io.knobs[index].inByte,
+                stopFocusSelectMode: function(effectGroup, unfocusGroup, noEffectFocused) {
+                    this.group = noEffectFocused ? unfocusGroup : effectGroup;
+                    this.inKey = noEffectFocused ? "meta" : `parameter${index + 1}`;
+                    this.shift = noEffectFocused ? undefined : function() {
+                        this.setGroupKey(unfocusGroup, "meta");
+                    };
+                    this.unshift = noEffectFocused ? undefined : function() {
+                        this.setGroupKey(effectGroup, `parameter${  index + 1}`);
+                    };
+                }
             });
             this.buttons[index] = new Button({
                 unit: this,
+                index: index,
```

Comment:
This doesn't seem used. Remove?

Reply:
it's used for
`this.unit.setFocusedEffect(index)`
a few lines below